source/common/router/router.cc

```diff
@@ -850,16 +883,17 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
                    (!active_shadow_policies_.empty() && !streaming_shadows_) ||
                    (route_entry_ && route_entry_->internalRedirectPolicy().enabled());
   if (buffering &&
-      getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_) {
+      ((!increased_retry_shadow_buffer_limit_until_got_connection_ &&
+        getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_) ||
```

Comment:
Let's take a step back for your first [commit #1](https://github.com/envoyproxy/envoy/pull/32359/commits/29a8867730776f031caec097e295e74e72d78998) with explicitly calling `onDecoderFilterAboveWriteBufferHighWatermark`.

I am actually looking into how we are buffering right now from this code path, taking a look at this comment: https://github.com/envoyproxy/envoy/blob/b47fc6648d7c2dfe0093a601d44cb704b7bad382/source/common/router/router.cc#L901-L907, it states we don't need to worry about the 413 since we early give up if the data is over-limit here. 

I doubt we probably can continue buffer the data if `getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_)` once, and it is actually using streaming mode with  `state_.decoder_filters_streaming_ = true`, and the watermark can be applied without worrying 413 in filter_manager and `onDecoderFilterAboveWriteBufferHighWatermark` is called already when we are moving data to the `requests_buffered_data` IIUC
https://github.com/envoyproxy/envoy/blob/b47fc6648d7c2dfe0093a601d44cb704b7bad382/source/common/http/filter_manager.cc#L542-L547, this is just like how watermark buffer works: "add data at first and then check the watermark".  cc @mattklein123 who added this. 

In this case we don't need to adjust the downstream decoder buffer limit, but simply to let this code go through when we are waiting for connection.

@fl0Lec WDYT and do you wanna give this a shot?

Reply:
> Sorry was OOO,
> 
> I would also rely on the buffer reaching the watermark to pause the read. So:
> 
> * getting connection from downstream
> * trying to get a connection with upstream
> * filter buffer gets full (we go over the watermark which pauses the downstream read)
> * we get a connection
> 
> At that point to go bellow the watermark, we have the following options:
> 
> 1. draining the buffer: no possible as other filter might also use it.
> 2. increasing the limit to be right above the currently buffered data. Probably the option for now
> 3. do we have other option ?
> 
> WDYT @botengyao ?

What is the specific reason that we cannot drain the buffer since router is the terminated filter?

Note, we may cannot drain here if we have shadow policy since the data will also be used for shadow, the draining should be conditional. Eventually this can be drained if router will not use the data anymore/already shipped to upstream, but this looks like not today's case. cc @alyssawilk.

Reply:
thanks, then we need to re-adjust the buffer limit with the current buffer length with a hard limit to drop with this approach. This delays the adjustment of buffer limit but still needs to adjust.

Reply:
That makes sense, I implemented it in the latest commit (0828ffd19616a2cf02903ba4426fbd2aa93d0453). Let's see if all test pass